### PR TITLE
Add a mechanism for automatically generating the detailed description of each cnf test.

### PR DESCRIPTION
**Pull request by fedepaol**
We want to provide a description of the tests. At the same time, we want to take track if new undocumented tests are added.
Here we provide a way to generate (and extend) a map between the test names and their (manually fed) descriptions.
Then, we use that very same file to generate the markdown doc using a template.

Finally, when an uncommented test is found, the fill command fails. This is tied to the CI, to avoid committing new uncommented tests.

### CODE REVIEW

**Comment by fedepaol** [comment]
/hold
Still need to fill the descriptions

**Comment by fedepaol** [comment]
/hold cancel
Filled out all the descriptions

File: cnf-tests/TESTLIST.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
+
+<!--- IMPORTANT!
+This file is generated manually. To add a new description please run run
```
Comment (SchSeba) [inline cnf-tests/TESTLIST.md:3]: run run?

File: cnf-tests/docgen/cmd/generate.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+var toClean = [](*regexp.Regexp){
+	regexp.MustCompile("\\[rfe_id:\\w+\\]"),
+	regexp.MustCompile("\\[crit:\\w+\\]"),
+	regexp.MustCompile("\\[level:\\w+\\]"),
+	regexp.MustCompile("\\[test_id:\\d+\\]"),
+	regexp.MustCompile("\\[vendor:cnf-qe@redhat.com\\]"),
+}
+
+var testListTemplate = `
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:28]: nic: is better to use `const` here

File: cnf-tests/docgen/cmd/generate.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+| Test Name | Description |
+| -- | ----------- |{{ with .OthersList }}{{ range . }}
+| {{ .Name }} | {{ .Description }} | {{ end }}{{ end }}
+`
+
+// generateCmd represents the generate command
+var generateCmd = &cobra.Command{
+	Use:   "generate",
+	Short: "A brief description of your command",
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:84]: can you edit this?

File: cnf-tests/docgen/cmd/generate.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	Long: `A longer description that spans multiple lines and likely contains examples
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:85]: can you edit this?

File: cnf-tests/docgen/cmd/generate.go (hunk 2 above)
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:28]: Can we move this to a template go file?

File: cnf-tests/docgen/cmd/generate.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	rootCmd.AddCommand(generateCmd)
+	generateCmd.Flags().StringVar(&markdownFile, "target", "", "The path of the markdown file")
+	generateCmd.Flags().StringVar(&e2eDescriptions, "testsjson", "", "The json file containing the descriptions of the e2e tests")
+	generateCmd.Flags().StringVar(&validationDescriptions, "validationjson", "", "The json file containing the descriptions of the validation tests")
+	cobra.MarkFlagRequired(generateCmd.LocalFlags(), "target")
+	cobra.MarkFlagRequired(generateCmd.LocalFlags(), "testsjson")
+	cobra.MarkFlagRequired(generateCmd.LocalFlags(), "validationjson")
+}
+
+type TestDescription struct {
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:106]: Can we move this to a template go file?

File: cnf-tests/docgen/cmd/generate.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	Name        string `json:"name"`
+	Description string `json:"description"`
+}
+
+type TemplateData struct {
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:111]: Can we move this to a template go file?

Comment (SchSeba) [inline cnf-tests/docgen/cmd/generate.go:111]: can this be just a list of `[]TestDescription` then extend the `TestDescription` struct to something like

```golang
suiteName string
tests []TestDescription
```

This way we can make the template much smaller, just have two for loop one over suites and inside one over the tests

File: cnf-tests/docgen/cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+	"github.com/spf13/cobra"
+)
+
+var cfgFile string
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "docs",
+	Short: "Cnf tests doc generator",
```
Comment (SchSeba) [inline cnf-tests/docgen/cmd/root.go:18]: Cnf -> CNF

File: hack/fill-empty-docs.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+export SUITES_PATH=cnf-tests/bin
+
+# generate the junit files
+hack/build-test-bin.sh
+cnf-tests/test-run.sh -junit _cache/junit -ginkgo.dryRun -ginkgo.v
+
+go build -o _cache/docgen cnf-tests/docgen/main.go 
+# use the junit files to fill the descriptions
+_cache/docgen fill --junit _cache/junit/cnftests-junit.xml --description cnf-tests/docgen/e2e.json
+_cache/docgen fill --junit _cache/junit/validation_junit.xml --description cnf-tests/docgen/validation.json
```
Comment (SchSeba) [inline hack/fill-empty-docs.sh:17]: nit: why this file is `_` and the other one is `-`?

File: hack/fill-empty-docs.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+set -e
+
+GOPATH="${GOPATH:-~/go}"
```
Comment (SchSeba) [inline hack/fill-empty-docs.sh:4]: can you put this in a common.sh and reuse it?

File: hack/fill-empty-docs.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+export GOFLAGS="${GOFLAGS:-"-mod=vendor"}"
+
+export PATH=$PATH:$GOPATH/bin
+export SUITES_PATH=cnf-tests/bin
+
+# generate the junit files
+hack/build-test-bin.sh
+cnf-tests/test-run.sh -junit _cache/junit -ginkgo.dryRun -ginkgo.v
```
Comment (SchSeba) [inline hack/fill-empty-docs.sh:12]: can you add the `_cache` in the gitignore?

**Review by SchSeba — COMMENTED** [review]
Great PR!

I have some comments

File: cnf-tests/TESTLIST.md (hunk 1 above)
Comment (fedepaol) [reply cnf-tests/TESTLIST.md:3]: My bad, need to fix.

**Comment by SchSeba** [comment]
@fedepaol also we need to open a PR against the release repo (create a new lane to fail if a new test exists and not documented).

Also if you can add this to the README file for anyone that wants to contribute to the project

**Comment by fedepaol** [comment]
> @fedepaol also we need to open a PR against the release repo (create a new lane to fail if a new test exists and not documented).
> 

It's already there, just under the make ci command that is currently called by CI.

> Also if you can add this to the README file for anyone that wants to contribute to the project

Yep, will do.

File: cnf-tests/docgen/cmd/generate.go (hunk 6 above)
Comment (fedepaol) [reply cnf-tests/docgen/cmd/generate.go:111]: Makes sense, need to keep the validation separate though as the template is slightly different with a different description there.

File: hack/fill-empty-docs.sh (hunk 8 above)
Comment (fedepaol) [reply hack/fill-empty-docs.sh:17]: It ended up being like this, not changed in this PR. That could be made better.

File: hack/fill-empty-docs.sh (hunk 10 above)
Comment (fedepaol) [reply hack/fill-empty-docs.sh:12]: It's already there

File: hack/common.sh [hunk 11]
```diff
@@ -9,3 +9,8 @@ function finish {
     popd
 }
 trap finish EXIT
+
+GOPATH="${GOPATH:-~/go}"
+export GOFLAGS="${GOFLAGS:-"-mod=vendor"}"
+
+export PATH=$PATH:$GOPATH/bin
```
Comment (SchSeba) [inline hack/common.sh:16]: nit: new line here please

File: Makefile [hunk 12]
```diff
@@ -71,3 +71,10 @@ cnf-tests-local:
 	$(IMAGE_BUILD_CMD) build --no-cache -f cnf-tests/Dockerfile -t cnf-tests-local .
 	$(IMAGE_BUILD_CMD) build --no-cache -f tools/s2i-dpdk/Dockerfile -t dpdk tools/s2i-dpdk/
 
+check-tests-nodesc:
```
[the remark this answers is not in the document]
Comment (fedepaol) [reply Makefile:74]: Apologies, I was sure I did it!

File: hack/common.sh (hunk 11 above)
Comment (fedepaol) [reply hack/common.sh]: Done

**Review by SchSeba — COMMENTED** [review]
/lgtm

**Pull request merged**